feat(fileLoader): Add support for loading buffer

algomaster99 opened the pull request:
Signed-off-by: Aman Sharma <mannu.poski10@gmail.com>

# Issue #741 
Add support for loading `Buffer` type content.

### Changes
- Add an argument `buffer` to specify if the file needs to be parsed as a `Buffer` type.
- Add test for checking if `logo.png` is loaded as buffer.

### Flags
- I haven't made another function `loadFromBuffer` as suggested in the issue. I felt it was possible by adding a few lines in the existing function itself.

algomaster99 commented:
@jeromesimeon 
I think the tests are not able to find the `logo.png` whose path I have hard coded in the tests. Can you help me fix it?

jeromesimeon commented on packages/ergo-compiler/lib/fileloader.js:
```diff
@@ -90,16 +90,21 @@ class FileLoader {
      * @param {string} fileName the relative file name
      * @param {boolean} json if true the file is converted to a JS Object using JSON.parse
      * @param {boolean} required whether the file is required
+     * @param {Buffer} buffer if true the file is converted to a Buffer
```
This is inconsistent with the type of the parameter (boolean).

jeromesimeon commented on packages/ergo-compiler/lib/fileloader.js:
```diff
@@ -90,16 +90,21 @@ class FileLoader {
      * required is false
      */
-    static async loadFileContents(path, fileName, json=false, required=false) {
+    static async loadFileContents(path, fileName, json=false, required=false, buffer=false) {
```
I think I'd rather have a separate `loadFileBuffer` method, which wouldn't need a `json` parameter (which is ignored in this case). It might be useful to have a way to cap the size of the buffer (throwing an error for files above a certain size?).

algomaster99 replied:
@jeromesimeon What size of image do you recommend? How many KBs approx?

jeromesimeon replied:
> @jeromesimeon What size of image do you recommend? How many KBs approx?

I'm not sure. I would raise this question in the issue and Slack Cicero channel.

jeromesimeon replied:
What's a typical size for a 256x256 image in format XXX?

jeromesimeon commented on packages/ergo-compiler/test/fileloader.js:
```diff
@@ -0,0 +1,31 @@
+
+    describe('#loadFileContents', () => {
+        it('should return a Buffer if buffer is true', async () => {
+            const content = await FileLoader.loadFileContents('packages/ergo-compiler/test/data', 'logo.png', false, false, true);
```
The path should be relative to the package. So `./test/data` should be the directory as in https://github.com/accordproject/ergo/blob/c452d1a7b80c75e0723c918148f15cb2c2fa1a42/packages/ergo-compiler/test/logicmanager.js#L33

algomaster99 commented:
@jeromesimeon 
- I did create the `loadZipFileBuffer` but couldn't write tests for it. I need a way to mock it by creating a `.cta` file with `logo.png`.
- Restricting the size of the image is yet to be to done. I shall use [this package](https://github.com/lovell/sharp) as suggested by you.

algomaster99 commented:
@jeromesimeon I believe using https://github.com/lovell/sharp for image dimension calculations can be an overkill. I suggest using [this package](https://www.npmjs.com/package/image-size) instead.
And for validating images as `PNG` only, we can use https://github.com/sindresorhus/image-type.

Here are some bundle size comparisons:
| Package | Bundle Size |
|:-:|:-:|
| [sharp](https://www.npmjs.com/package/sharp)  | Unpacked Size - 396kB  |
| [image-size](https://www.npmjs.com/package/image-size)  |  Minified - 14kB  |
| [image-type](https://github.com/sindresorhus/image-type)  | Minified - 10.7kB  |
> Source https://bundlephobia.com/

algomaster99 commented:
@jeromesimeon I have added constraints for checking image dimensions (using `image-size`) and mime-type (using `image-type`).
- I didn't restrict it to `256x256` because I feel the user should give any square image and we should resize it.
- I have added two more parameters `aspectRatio` and `epsilon` because I wanted to give users some freedom and they don't *necessarily* have to upload an image with strict dimensions. `epsilon` is the allowance a user can be given.
- For `mime-type`, only `image/png` is allowed for now. But is extensible by expanding `ALLOWED_MIMETYPES` array.
- I also haven't restricted the user to name the image `logo.png` as we should be the one to save it as `logo.png`.

algomaster99 commented:
@jeromesimeon no problem at all. But I have a few questions first.
- so in `ergo` it should just load the buffer without checking any constraints, right?
- all the constraint checking such as (dimensions and mime type) should happen in `cicero`

Am I right? If not, please correct me.

jeromesimeon commented:
> @jeromesimeon no problem at all. But I have a few questions first.
> 
> * so in `ergo` it should just load the buffer without checking any constraints, right?
> * all the constraint checking such as (dimensions and mime type) should happen in `cicero`
> 
> Am I right? If not, please correct me.

That's my suggestion. _if_ eventually we want to check constraints on the buffer those would be about the buffer, not assuming we know whether it's pdf or image or any other kind of binary (in Ergo).

jeromesimeon commented:
> @jeromesimeon I have added constraints for checking image dimensions (using `image-size`) and mime-type (using `image-type`).
> 
> * I didn't restrict it to `256x256` because I feel the user should give any square image and we should resize it.

It seems reasonable, but I am a little bit uncertain. When you resize automatically you may lose quality and some people might want to apply their own process to resize. I'm also wondering if we should allow a few variant sizes.

I think this discussion could use a bit of a UX designer input @Michael-Grover ?

> * I have added two more parameters `aspectRatio` and `epsilon` because I wanted to give users some freedom and they don't _necessarily_ have to upload an image with strict dimensions. `epsilon` is the allowance a user can be given.

I like this.

> * For `mime-type`, only `image/png` is allowed for now. But is extensible by expanding `ALLOWED_MIMETYPES` array.

That seems very reasonable to me.

> * I also haven't restricted the user to name the image `logo.png` as we should be the one to save it as `logo.png`.

I think I disagree. How will you know where the logo is when it's in a template directory (i.e., before you create the archive)?

dselman commented:
I'm in favour of keeping it simple and quite strict on image format and size. We want these logos to be small and used to represent the template in a list of templates: for example in the Template Library cicero-ui React component. We can always relax constraints in the future - much harder to tighten them.

I'm also reluctant to add 3rd-party libs, if we can avoid doing so. Every library we add makes the bundle size larger and is more maintenance to continually update the library version to incorporate security fixes etc.

jeromesimeon commented:
> I'm also reluctant to add 3rd-party libs, if we can avoid doing so. Every library we add makes the bundle size larger and is more maintenance to continually update the library version to incorporate security fixes etc.

Do you know of a way to check an image type and size without a library?

dselman commented:
Size we could check based on the byte[] - or we need to go deeper into the PNG (at which point we might as well use a 3rd-party).

Here's some simple code to check the PNG file header.
https://github.com/arian/pngjs/blob/c1a2f5852686a22229f5f3154f974011162b5263/PNGReader.js#L80

jeromesimeon commented:
> Size we could check based on the byte[] - or we need to go deeper into the PNG (at which point we might as well use a 3rd-party).
> 
> Here's some simple code to check the PNG file header.
> https://github.com/arian/pngjs/blob/c1a2f5852686a22229f5f3154f974011162b5263/PNGReader.js#L80

I'm finding it harder to determine how to now the length and height of the image at the moment.

jeromesimeon commented:
I _think_ this is how `image-size` determines the height and width:
https://github.com/image-size/image-size/blob/a2531b1c5bb8e93d0836a693962f54ffdc5f0af8/lib/types/png.ts#L24

I don't really understand it yet.
